### PR TITLE
CIRC-886 Truncate due date based on patron expiration at check out

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Approach
- Add truncate due date logic if patron expires earlier for check-out;
- Create tests to cover all possible scenarious;

Resolves: [CIRC-886](https://issues.folio.org/browse/CIRC-886)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 1]
```diff
@@ -1906,6 +1919,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
     assertThat(loan.getString("action"), is(CHECKED_OUT_THROUGH_OVERRIDE));
   }
 
+  @Test
+  public void dueDateShouldBeTruncatedToTheEndOfLastWorkingDayBeforePatronExpiration() {
```
Comment (roman-barannyk) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1923]:
```suggestion
  public void dueDateShouldBeTruncatedToTheEndOfPreviousOpenDayBeforePatronExpiration() {
```

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 2]
```diff
@@ -1906,6 +1919,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+        .forItem(item)
+        .to(steve)
+        .at(CASE_ONE_DAY_IS_OPEN_NEXT_TWO_DAYS_CLOSED)).getJson();
+
+    assertThat(DateTime.parse(response.getString("dueDate")).toDateTime(),
+      is(FIRST_DAY_OPEN.toDateTime(LocalTime.MIDNIGHT.minusSeconds(1), UTC)));
+  }
+
+  @Test
+  public void dueDateShouldBeTruncatedToTheEndOfNextWorkingDayAfterPatronExpiration() {
```
Comment (roman-barannyk) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1940]:
```suggestion
  public void dueDateShouldBeTruncatedToTheEndOfNextOpenDayAfterPatronExpiration() {
```

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 3]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+    use(buildLoanPolicyWithFixedLoan(MOVE_TO_THE_END_OF_THE_PREVIOUS_OPEN_DAY));
+
+    IndividualResource item = itemsFixture.basedUponNod();
+    IndividualResource steve = usersFixture.steve(user -> user.expires(DateTime.now().plusDays(3)));
+
+    JsonObject response = checkOutFixture.checkOutByBarcode(
+      new CheckOutByBarcodeRequestBuilder()
+        .forItem(item)
+        .to(steve)
+        .at(CASE_ONE_DAY_IS_OPEN_NEXT_TWO_DAYS_CLOSED)).getJson();
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1944]: Rename constant

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 4]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
     assertThat(loan.getString("action"), is(CHECKED_OUT_THROUGH_OVERRIDE));
   }
 
+  @Test
+  public void dueDateShouldBeTruncatedToTheEndOfPreviousOpenDayBeforePatronExpiration() {
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1934]: Make closed library due date management to apply in this case

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 5]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+    assertThat(DateTime.parse(response.getString("dueDate")).toDateTime(),
+      is(YESTERDAY_OPEN_DAY.toDateTime(LocalTime.MIDNIGHT.minusSeconds(1), UTC)));
+  }
+
+  @Test
+  public void dueDateShouldBeTruncatedToThePatronsExpirationDateTimeIfKeepCurrentDueDateStrategy() {
+    use(buildLoanPolicyWithFixedLoan(KEEP_THE_CURRENT_DUE_DATE));
+
+    IndividualResource item = itemsFixture.basedUponNod();
+    DateTime patronExpirationDate = DateTime.now().plusDays(3).withZone(UTC);
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1974]: Change expiration time

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 6]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+    IndividualResource steve = usersFixture.steve(user -> user.expires(patronExpirationDate));
+
+    JsonObject response = checkOutFixture.checkOutByBarcode(
+      new CheckOutByBarcodeRequestBuilder()
+        .forItem(item)
+        .to(steve)
+        .at(CASE_ONE_DAY_IS_OPEN_NEXT_TWO_DAYS_CLOSED)).getJson();
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1981]: We need a service point with empty calendar

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 7]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+
+    IndividualResource item = itemsFixture.basedUponNod();
+    DateTime patronExpirationDate = DateTime.now().plusHours(1).withZone(UTC);
+    IndividualResource steve = usersFixture.steve(user -> user.expires(patronExpirationDate));
+
+    JsonObject response = checkOutFixture.checkOutByBarcode(
+      new CheckOutByBarcodeRequestBuilder()
+        .forItem(item)
+        .to(steve)
+        .at(CASE_ONE_DAY_IS_OPEN_NEXT_TWO_DAYS_CLOSED)).getJson();
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1998]: Same here

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 8]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+        .to(steve)
+        .at(CASE_YESTERDAY_OPEN_TODAY_CLOSED_TOMORROW_OPEN)).getJson();
+
+    assertThat(DateTime.parse(response.getString("dueDate")).toDateTime(),
+      is(YESTERDAY_OPEN_DAY.toDateTime(END_TIME_SECOND_PERIOD, UTC)));
+  }
+
+  @Test
+  public void
+  dueDateShouldBeTruncatedToTheEndOfPreviousServicePointHoursIfMoveToTheBeginningOfNextStrategy() {
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:2024]: Make sure it doesn't fail later in the day, when there is no "next open hours" period

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 9]
```diff
@@ -1916,6 +1930,127 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+        .at(CASE_YESTERDAY_OPEN_TODAY_CLOSED_TOMORROW_OPEN)).getJson();
+
+    assertThat(DateTime.parse(response.getString("dueDate")).toDateTime(),
+      is(YESTERDAY_OPEN_DAY.toDateTime(END_TIME_SECOND_PERIOD, UTC)));
+  }
+
+  private LoanPolicyBuilder buildLoanPolicyWithFixedLoan(DueDateManagement strategy) {
+    return new LoanPolicyBuilder()
+      .fixed(loanPoliciesFixture.createExampleFixedDueDateSchedule().getId())
+      .withLoanPeriod(Period.months(1))
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:2044]: Should it have loan period of 1 month?

File: src/main/java/org/folio/circulation/domain/policy/library/EndOfPreviousOpenHoursTruncateStrategy.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  private final DateTime startDateTime;
+
+  public EndOfPreviousOpenHoursTruncateStrategy(DateTime startDateTime, DateTimeZone zone) {
+    super(zone);
+    this.startDateTime = startDateTime;
+  }
+
+  @Override
+  public Result<DateTime> calculateDueDate(DateTime requestedDate, AdjacentOpeningDays openingDays) {
+    Objects.requireNonNull(openingDays);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/policy/library/EndOfPreviousOpenHoursTruncateStrategy.java:25]: Check that this exception is handled correctly

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 11]
```diff
@@ -1916,6 +1930,162 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+  @Test
+  public void dueDateShouldBeTruncatedToTheEndOfPreviousOpenDayBeforePatronExpiration() {
+    DateTime loanDate = SECOND_DAY_CLOSED.toDateTime(LocalTime.MIDNIGHT.plusHours(10));
+    use(buildLoanPolicyWithFixedLoan(MOVE_TO_THE_END_OF_THE_PREVIOUS_OPEN_DAY, loanDate.plusDays(1)));
+
+    IndividualResource item = itemsFixture.basedUponNod();
+    DateTime patronExpirationDate = loanDate.plusHours(2);
+    IndividualResource steve = usersFixture.steve(user -> user.expires(patronExpirationDate));
+
+    JsonObject response = executeAtFixedTime(() -> checkOutFixture.checkOutByBarcode(
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1942]: executeWithFixedDateTime

File: src/test/java/api/loans/CheckOutByBarcodeTests.java [hunk 12]
```diff
@@ -1916,6 +1930,162 @@ public void canOverrideManualAndAutomationPatronBlocksWhenBlocksArePresent() {
...
+    IndividualResource item = itemsFixture.basedUponNod();
+    DateTime patronExpirationDate = loanDate.plusHours(1);
+    IndividualResource steve = usersFixture.steve(user -> user.expires(patronExpirationDate));
+
+    JsonObject response = executeAtFixedTime(() -> checkOutFixture.checkOutByBarcode(
+      new CheckOutByBarcodeRequestBuilder()
+        .forItem(item)
+        .to(steve)
+        .on(loanDate)
+        .at(CASE_FIRST_DAY_OPEN_SECOND_CLOSED_THIRD_OPEN)).getJson(), loanDate);
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckOutByBarcodeTests.java:1967]: Should the third day be closed here?

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**